File: 46-permutations/46-permutations.cpp
class Solution {
public:
    
    void findPermutations(vector<int>& nums, vector<vector<int>>& solution, int index){
        
        if(index==nums.size()){
            solution.push_back(nums);    
            return;
        }
        
        
        int n = nums.size();
        for(int i=index;i<n;i++){
            
            swap(nums[index],nums[i]);
            findPermutations(nums, solution, index+1);
            swap(nums[index],nums[i]);
        
        }
        
        
    }
    
    
    
    vector<vector<int>> permute(vector<int>& nums) {
        
        vector<vector<int>> solution;
        findPermutations(nums,solution,0);
        return solution;
    
    }
};

File: 46-permutations/46-permutations.py
def findPermutation(pos, numsLength, nums, res):
        
        if(pos==numsLength):
            res.append(nums[:])
        else:
            for idx in range(pos,numsLength):
                nums[pos], nums[idx] = nums[idx], nums[pos]
                findPermutation(pos+1,numsLength,nums,res)
                nums[pos], nums[idx] = nums[idx], nums[pos]

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        res = []
        pos=0
        numsLength = len(nums)        
        findPermutation(pos,numsLength,nums,res)
        return res